components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java

```diff
@@ -79,79 +81,91 @@ public UndertowEndpoint getEndpoint() {
-            request.setPath(pathAndQuery);
-            request.setMethod(method);
+        final HeaderMap requestHeaders = request.getRequestHeaders();
 
-            final HeaderMap requestHeaders = request.getRequestHeaders();
+        // Set the Host header
+        final Message message = camelExchange.getIn();
+        final String host = message.getHeader(Headers.HOST_STRING, String.class);
+        requestHeaders.put(Headers.HOST, Optional.ofNullable(host).orElseGet(() -> uri.getAuthority()));
 
-            // Set the Host header
-            Message message = exchange.getIn();
-            final String host = message.getHeader("Host", String.class);
-            requestHeaders.put(Headers.HOST, Optional.ofNullable(host).orElseGet(()-> uri.getAuthority()));
+        final Object body = undertowHttpBinding.toHttpRequest(request, camelExchange.getIn());
 
-            Object body = getRequestBody(request, exchange);
+        final TypeConverter tc = endpoint.getCamelContext().getTypeConverter();
+        final ByteBuffer bodyAsByte = tc.tryConvertTo(ByteBuffer.class, body);
 
-            TypeConverter tc = endpoint.getCamelContext().getTypeConverter();
-            ByteBuffer bodyAsByte = tc.tryConvertTo(ByteBuffer.class, body);
+        if (body != null) {
+            requestHeaders.put(Headers.CONTENT_LENGTH, bodyAsByte.remaining());
+        }
 
-            if (body != null) {
-                requestHeaders.put(Headers.CONTENT_LENGTH, bodyAsByte.array().length);
-            }
+        for (final Map.Entry<String, List<String>> entry : cookieHeaders.entrySet()) {
+            requestHeaders.putAll(HttpString.tryFromString(entry.getKey()), entry.getValue());
+        }
 
-            if (getEndpoint().getCookieHandler() != null) {
-                Map<String, List<String>> cookieHeaders = getEndpoint().getCookieHandler().loadCookies(exchange, uri);
-                for (Map.Entry<String, List<String>> entry : cookieHeaders.entrySet()) {
-                    requestHeaders.putAll(new HttpString(entry.getKey()), entry.getValue());
-                }
-            }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Executing http {} method: {}", method, pathAndQuery);
+        }
 
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Executing http {} method: {}", method, pathAndQuery);
-            }
-            connection = connect.get();
-            connection.sendRequest(request, new UndertowProducerCallback(connection, bodyAsByte, exchange, callback));
+        final UndertowClientCallback clientCallback = new UndertowClientCallback(camelExchange, callback, getEndpoint(),
+            request, bodyAsByte);
 
-        } catch (Exception e) {
-            IOHelper.close(connection);
-            exchange.setException(e);
-            callback.done(true);
-            return true;
-        }
+        client.connect(clientCallback, uri, worker, ssl, pool, options);
```

Comment:
Ah so connect, will then have undertow signal completed to the connect event, where you then send the request and listen for the response and all that stuff which is in that UndetowClientCallback. If so maybe add a code comment here about this, so we would know that only calling connect is not missing something else.